### PR TITLE
Add an expiration to card grants

### PR DESCRIPTION
**Pull request by sampoder**
Default to 2 years. Organizers can change it.

### CODE REVIEW

File: app/jobs/card_grant_expiration_job.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class CardGrantExpirationJob < ApplicationJob
+  queue_as :low
+  def perform
+    CardGrant.joins(:card_grant_setting).where("card_grants.created_at + (card_grant_settings.expiration_preference * interval '1 day') < NOW()").find_each { |card_grant|
+      card_grant.cancel!(User.find_by(email: "bank@hackclub.com"), expired: true)
+    }
```
Comment (garyhtou) [inline app/jobs/card_grant_expiration_job.rb]: When we have a multiline block, please use the `do` keyword

File: app/jobs/card_grant_expiration_job.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  end
```
Comment (garyhtou) [inline app/jobs/card_grant_expiration_job.rb]:
```suggestion
  def perform
    hcb_user = User.find_by!(email: "bank@hackclub.com")
    CardGrant.joins(:card_grant_setting).where("card_grants.created_at + (card_grant_settings.expiration_preference * interval '1 day') < NOW()").find_each do |card_grant|
      card_grant.cancel!(hcb_user, expired: true)
    end
  end
```

File: app/models/card_grant.rb [hunk 3]
```diff
@@ -139,7 +141,7 @@ def cancel!(canceled_by)
 
     end
 
-    update!(status: :canceled)
+    update!(status: :expired)
```
Comment (garyhtou) [inline app/models/card_grant.rb]: We add logic to keep the cancelled state

File: app/models/card_grant.rb [hunk 4]
```diff
@@ -121,9 +123,9 @@ def topup!(amount_cents:, topped_up_by: User.find(sent_by_id))
     end
   end
 
-  def cancel!(canceled_by)
+  def cancel!(canceled_by, expired: false)
```
Comment (garyhtou) [inline app/models/card_grant.rb]: I would recommend

```ruby
def expire!
  hcb_user = User.find_by!(email: "hcb@hackclub.com")
  cancel!(hcb_user, expire: true)
end
```

File: db/migrate/20240819163511_add_expiration_to_card_grant_setting.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddExpirationToCardGrantSetting < ActiveRecord::Migration[7.1]
+  def change
+    add_column :card_grant_settings, :expiration_preference, :integer, null: false, default: 365 * 2 # days
```
Comment (garyhtou) [inline db/migrate/20240819163511_add_expiration_to_card_grant_setting.rb]: Let's make default to 1 year

File: app/models/card_grant_setting.rb [hunk 6]
```diff
@@ -26,4 +27,11 @@ class CardGrantSetting < ApplicationRecord
   alias_attribute :allowed_categories, :category_lock
   has_many :card_grants, through: :event
 
+  enum :expiration_preference, {
+    "90 days": 90,
+    "6 months": 183,
+    "1 year": 365,
+    "2 years": 365 * 2
+  }
```
Comment (garyhtou) [inline app/models/card_grant_setting.rb]: Is there a way for Rails to define the type? Can we have it typecast to days?

Comment (garyhtou) [inline app/models/card_grant_setting.rb]:
```suggestion
  }, prefix: :expires_after
```

File: app/views/card_grants/_spending_instructions.html.erb [hunk 7]
```diff
@@ -40,7 +40,11 @@
           <strong>Billing address</strong>
           <%= render_address stripe_card.cardholder %>
         </p>
-
+      </section>
+      <section class="card__banner card__banner--bottom card__darker pt2 pb2">
+        <p class="muted mt0 mb0">
+          The card will be available for use until <%= format_date card_grant.expires_on %>.
```
Comment (garyhtou) [inline app/views/card_grants/_spending_instructions.html.erb]:
```suggestion
          The card will be available for use until <%= format_date card_grant.expires_on %>. Afterward, the remaining funds will be returned to <%= card_grant.event.name %>.
```

File: app/jobs/card_grant_expiration_job.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class CardGrantExpirationJob < ApplicationJob
+  queue_as :low
+  def perform
+    CardGrant.joins(:card_grant_setting).where("card_grants.created_at + (card_grant_settings.expiration_preference * interval '1 day') < NOW()").find_each { |card_grant|
```
Comment (garyhtou) [inline app/jobs/card_grant_expiration_job.rb]:
```suggestion
    CardGrant.active.joins(:card_grant_setting).where("card_grants.created_at + (card_grant_settings.expiration_preference * interval '1 day') < NOW()").find_each { |card_grant|
```

**Review by garyhtou — DISMISSED** [review]
If you wanna get fancy, send a mailer to the organizer and maybe grant recipient

**Comment by garyhtou** [comment]
If you want to get even fanicer, maybe an activity

File: app/jobs/card_grant_expiration_job.rb (hunk 1 above)
Comment (sampoder) [reply app/jobs/card_grant_expiration_job.rb]:
```suggestion
    CardGrant.joins(:card_grant_setting).where("card_grants.created_at + (card_grant_settings.expiration_preference * interval '1 day') < NOW()").find_each do |card_grant|
      card_grant.cancel!(User.find_by(email: "bank@hackclub.com"), expired: true)
    end
```

File: db/migrate/20240819163511_add_expiration_to_card_grant_setting.rb (hunk 5 above)
Comment (sampoder) [reply db/migrate/20240819163511_add_expiration_to_card_grant_setting.rb]:
```suggestion
    add_column :card_grant_settings, :expiration_preference, :integer, null: false, default: 365 # days
```

File: db/schema.rb [hunk 9]
```diff
@@ -420,6 +420,7 @@
     t.string "merchant_lock"
     t.string "category_lock"
     t.string "invite_message"
+    t.integer "expiration_preference", default: 730, null: false
```
Comment (sampoder) [inline db/schema.rb]:
```suggestion
    t.integer "expiration_preference", default: 365, null: false
```

**Pull request merged**